lib/elastic_apm/transport/base.rb

```diff
@@ -13,40 +13,53 @@ class Base
 
       def initialize(config)
         @config = config
-        @connection = Connection.new(config)
-        @worker = Worker.new(config, connection)
+        @queue = SizedQueue.new(2000) # TODO: Max?
+        @size = 4
+        @pool = Concurrent::FixedThreadPool.new(@size)
+        @workers = []
+
+        @size.times do |i|
+          @pool.post do
+            debug 'Booting worker'
+            worker = Worker.new(config, queue, name: i)
+            @workers << worker
+            worker.work_forever
+            debug "After work -- #{worker.name}"
+            @workers -= [worker]
+            debug '!!! WORKER COUNT %i', @workers.count
+          end
+        end
       end
 
-      attr_reader :config, :filters, :connection, :worker
+      attr_reader :config, :filters, :queue, :messages
 
       def submit(resource)
-        ensure_worker
-        worker.push(resource)
+        queue.push(resource)
```

Comment:
If the queue is full this is blocking until space is available, which means the thread could potentially block for a really long time if the server connection is dead. WDYT about using the non-blocking push and catching the exception that is caught?